torch/_dynamo/graph_region_tracker.py

```diff
@@ -0,0 +1,290 @@
+class BfsRegionIter:
+    def __init__(self, origin: Node) -> None:
+        self._cur: Tuple[Optional[str], Optional[Node]] = (None, origin)
+        self._queue: Deque[Tuple[Optional[str], Optional[Node]]] = deque()
+
+    @staticmethod
+    def create(origin: Node) -> "BfsRegionIter":
+        it = BfsRegionIter(origin)
+        it.add_children(origin)
+        return it
+
+    def next(self) -> Tuple[Optional[str], Optional[Node]]:
+        ret = self._cur
+        if not self._queue:
+            self._cur = (None, None)
+        else:
+            self._cur = self._queue.popleft()
+        return ret
+
+    def peek(self) -> Tuple[Optional[str], Optional[Node]]:
+        return self._cur
+
+    def add_children(self, node: Node) -> None:
+        arg: Any
+        for arg in node.args:
+            if isinstance(arg, Node):
+                self._append((None, arg))
+
+        key: str
+        kwarg: Any
+        for key, kwarg in node.kwargs.items():
+            if isinstance(kwarg, Node):
+                self._append((key, kwarg))
+
+    def _append(self, name_arg: Tuple[Optional[str], Node]) -> None:
+        if self._cur == (None, None):
+            self._cur = name_arg
+        else:
+            self._queue.append(name_arg)
+
+
+class GraphRegionTracker:
+    def __init__(self) -> None:
+        self.loc_to_duplicates: Dict[str, IdenticalNodes] = defaultdict(list)
+        self.node_to_duplicates: Dict[Node, IdenticalNodes] = {}
+        self.node_to_global_state_hash: Dict[Node, int] = {}
+        self.input_pickler = InputPickler()
+
+    def _hash_node(
+        self, filename: str, lineno: int, instruction_pointer: Optional[int], node: Node
+    ) -> str:
+        from torch._inductor.codecache import sha256_hash
+
+        key = (
+            get_global_state_key(),
+            filename,
+            lineno,
+            instruction_pointer,
+            _normalize_args(node),
+        )
+        return sha256_hash(self.input_pickler.dumps(key))
+
+    def track_node(self, tx: "InstructionTranslatorBase", node: Node) -> None:
+        duplicates = self.loc_to_duplicates[
+            self._hash_node(
+                tx.f_code.co_filename, tx.lineno, tx.instruction_pointer, node
+            )
+        ]
+        duplicates.append(node)
+        self.node_to_duplicates[node] = duplicates
+
+    def is_identical(self, n0: Node, n1: Node) -> bool:
+        return (
+            n0 in self.node_to_duplicates
+            and n1 in self.node_to_duplicates
+            and self.node_to_duplicates[n0] == self.node_to_duplicates[n1]
+            and n0 is not n1
+        )
+
+    def get_identical_regions(self, graph: torch.fx.Graph) -> List[List[Region]]:
+        topological_ranking = {node: i for i, node in enumerate(graph.nodes)}
+        group_ranking = {}
+        region_groups = []
+
+        # Create region groups; a region group is a group
+        # of regions that are all identical. In this initial state
```

Comment:
We don't really need to worry about this now, but hierarchical compilation (as currently implemented) has some limitations on what the regions can look like. Most notably, we require that the region is functional (no mutation or aliasing) at the boundaries. Have you thought about how to incorporate this into the region searching?

Reply:
Yeah I discussed this with Animesh a bit, we're going to either reuse the analysis in AOT or add support in invoke_subgraph for input aliasing, but we haven't reached a conclusion on this.